### DISCUSSION TITLE
Custom SMTP with Google Workspace

### DISCUSSION THREAD

**Discussion by ryan-eiv** [supabase/supabase#5320] [Questions] [answered]
Hello, 
I'm trying to set up a custom STMP for supabase auth with the Google Workspace STMP relay.

I've set up the google workspace gmail relay like this:
![image](https://user-images.githubusercontent.com/95147135/152545828-ba5c0de7-1ab1-426b-af3d-10ff4e7f7c16.png)

And I've set up my stmp on supabase like this:
![image](https://user-images.githubusercontent.com/95147135/152546009-486f378f-2e44-47ba-a5fd-34d2cc6af7b7.png)

(mygoogledomain.io is just a placeholder and that email account is a user inside Google Workspace)

However, I'm getting a 500 response from supabase when I call `signIn()`:
```
{message: "The upstream server is timing out"}
```

Is there a way to view the logs and figure out what the issue is?

**Comment by libovness** [comment]
@ryan-eiv I'm experiencing the same. Did you figure this out?

**Reply by ry-propeller** [reply]
It did work in the end. I don't remember what I did, but the reason why I wanted to set this up was to avoid login emails going to spam. Unfortunately, emails kept on going to spam, so we are now sending emails ourselves thru another email service and just getting the magic link on the backend

**Reply by FerasAloudah** [reply]
@libovness did you manage to solve this issue? I managed to get rid of the timeout error by changing the SMTP port to 25. However, I'm now getting an "Error sending magic link" error. Any thoughts?

Also, @ry-propeller how are you sending the magic links from your own backend? Are you self-hosting your Supabase instance?

**Reply by libovness** [reply]
@ry-propeller we have found that our emails are going to spam as well, even after switching to sendgrid. @FerasAloudah is this a known issue?

**Reply by FerasAloudah** [reply]
@libovness Hey, sorry for the late reply. We've been running into issues using SendGrid as well, we think this is related to our domain's reputation and should be fixed after a while. Asking your users to mark the email as not spam, could help speed this up for the time being.

**Reply by Verify21** [reply]
Hey guys, did any of you figure a workaround for preventing emails appearing in spam?

**Reply by libovness** [reply]
We did. I don't know if any of these were independently effective but for us using friendly HTML emails and following up with people who had auth emails land in their spam and encouraging them to report it as not spam seems to have done the trick. I think it's more of a brute force meatspace thing than a technical set up thing.

**Reply by vincelwt** [reply]
Also having all my signup emails going to spam with mailgun. Never had issues with my other mailgun domains previously. Will try setting gsuite smtp with an app password see if it works.

**Comment by libovness** [comment]
i managed to solve it by switching to sendgrid smtp

google smtp stopped working for reasons that remain unknown

**Reply by FerasAloudah** [reply]
Thanks for the quick reply! I'll give SendGrid a look.

**Comment by benlammers** [comment]
If anyone else runs into this issue with Gmail SMTP I solved mine by using a Google App Password not my accounts default password. This solved it for me!
https://support.google.com/accounts/answer/185833

**Reply by Andromed** [reply]
Hi @benlammers,

Could you let me know how did you manage to solve it with App Password ?

Thank you.

**Reply by benlammers** [reply]
I followed the instructions linked in my comment above to get an app password then I put the app password in for SMTP password in Supabase

<img width="677" alt="image" src="https://user-images.githubusercontent.com/41494387/176057733-d94a33f9-8fc3-430f-bb98-4a66feca1ef5.png">

**Reply by mfaywu** [reply]
Sadly it looks like app passwords are not supported by Google Workspace, I guess you can only use it with personal Gmail.

**Reply by alexbaramilis** [reply]
For me it worked with app password and Google Workspace.

**Reply by ffuyu** [reply]
App passwords can be used in Google Workspace too. You just need to enable 2FA.

**Reply by vaibhavmundra** [reply]
If you can't find the App Password setting in your Google Workspace account page, just do a search for it. It wasn't showing under the security settings in mine, but I found it via search.

<img width="1512" alt="Screenshot 2023-10-26 at 11 58 58 AM" src="https://github.com/supabase/supabase/assets/110548398/08fd214c-2275-4fc9-bf5c-d96e3c2073f2">

**Comment by itsondrej** [comment]
@benlammers @ryan-eiv Did you end up using `smtp-relay.gmail.com`? For us, it works with smtp.google.com (thanks for the app password trick btw), but we can't make it work with relay. Do we need to change some other parameters?

**Reply by benlammers** [reply]
Not familiar with relay. I am using the settings shown in the screenshot above

**Reply by alexbaramilis** [reply]
For me it worked with relay.

**Reply by alexbaramilis** [reply]
And now it stopped working. I'm getting AuthRetryableFetchError. Switched to smtp.gmail.com and seems to be working so far.

**Reply by ArianSha** [reply]
Hey @alexbaramilis did you continue to use smtp.gmail.com without any issues?

**Reply by alexbaramilis** [reply]
Hi @ArianSha, I was away for a few weeks and didn't check it but I just tested it now and it's still working.

**Reply by ArianSha** [reply]
Thanks for checking. Have you had any issues with emails being recognized as spam? I have the same setup, and some of my sign up mails is recognized as spam on outlook. I found a seperate thread about it, but none of these tips were helpful.

https://github.com/supabase/supabase/issues/4858

**Reply by alexbaramilis** [reply]
None so far, but it's not in production yet and I'm only testing with my personal (hotmail) and org email (gmail). I've read that marking an email as not spam helps, so you could try getting a notice to your users to check their spam folders and if they find the email, mark it as not spam.

**Comment by hnykda** [comment]
Using app password worked for me!

**Reply by dr-starchildren** [reply]
Was this with the smtp-relay.gmail.com or just the smtp.gmail.com? I've not had consistent success with either.

**Reply by hnykda** [reply]
smtp.gmail.com

**Comment by kjendrzyca** [comment]
The only combination that works for me is `smtp.gmail.com` and port `587`. Using anything else results in `AuthRetryableFetchError` and status code `504`.

**Reply by MigotMateusz** [reply]
Same for me. `smtp-relay.gmail.com` does not work at all for me

**Reply by kangmingtay** [reply]
Hey @kjendrzyca, port `465` should work too with `smtp.gmail.com`. Hopefully this helps, but here is my SMTP configuration:

![image](https://github.com/supabase/supabase/assets/28647601/7b71abef-9932-499a-baf2-9fd53c40a242)

![image](https://github.com/supabase/supabase/assets/28647601/283589c8-364c-4400-92af-1b436f43b185)

**Reply by libyandev** [reply]
it's work thank you

**Comment by kangmingtay** [comment, marked as answer]
Hey everyone, i've been hearing feedback about how challenging it can be to get google SMTP working properly with Supabase. I've tried setting this up with a trial google workspace account and this is what i've discovered.

1. The sender email and SMTP username has to be your google workspace admin email.
2. The SMTP password used has to be an app password. Google workspace doesn't make it easy to to figure out how to create this. You need to **enable 2FA** on your workspace account first (not inside the admin console but in https://myaccount.google.com/)
3. If you want to use `smtp-relay.gmail.com` only port 465 works. If you want to use `smtp.gmail.com`, you can use port 465 or 587 (both ports worked for me).

I'm also gonna include some screenshots below on what those steps look like:

## For `smtp-relay.gmail.com`
![image](https://github.com/supabase/supabase/assets/28647601/800448a1-0245-4a26-855f-3be8f44adf03)

(The "Require TLS encryption" option doesn't seem to affect anything during my testing)
![image](https://github.com/supabase/supabase/assets/28647601/5b55ddbc-abfc-4c66-a012-b8e4d23cd438)


## For `smtp.gmail.com`
![image](https://github.com/supabase/supabase/assets/28647601/ade24c85-50b5-47ca-a2ce-c3f77a0e128a)

**Reply by kangmingtay** [reply]
A lot of this was figured out through trial and error - i don't have explanations as to why only port 465 works with `smtp-relay.gmail.com` and / or why the TLS encryption option has no effect on SMTP. 

As to whether this is a Supabase Auth bug, i highly doubt that it is because we use the [native golang smtp library ](https://pkg.go.dev/net/smtp). If anyone still has issues getting google SMTP to work, i would recommend to switch to one of these SMTP providers:
* [Resend](https://resend.com/blog/how-to-configure-supabase-to-send-emails-from-your-domain) 
* [Sendgrid](https://sendgrid.com/en-us)
* [Mailgun](https://www.mailgun.com/)
* Good old AWS SES SMTP

**Reply by paulmiu** [reply]
Hi @kangmingtay,

Thanks for taking the time to look into this.

However, the whole point of using an SMTP relay is to *relay* emails for different email addresses. That is, if I want to send emails from hello@example.com, I don't want to create a new user account with that address in Google Workspace and give it admin privileges. This defeats the purpose of Google Workspace's SMTP relay.

If I want to send emails from my user account, I can just use the default Gmail SMTP server (`smtp.gmail.com`).

So the question remains:
How do I use the Google Workspace SMTP relay in Supabase to send emails from any email address within my domains associated with my Google Workspace account?

**Reply by alexbaramilis** [reply]
Thanks for the investigation. I can confirm the above.

smtp.gmail.com worked for me with 587 and 465 and smtp-relay.gmai.com only worked with 465.

The only thing I have to add is I'm not using my admin email and it still works with both setups. I'm using another address from my google workspace with 2FA and app password.

**Reply by kangmingtay** [reply]
> How do I use the Google Workspace SMTP relay in Supabase to send emails from any email address within my domains associated with my Google Workspace account?

@mamiu hmm, that makes sense, but it sounds like more of question for Google Workspace because it seems like it depends on how you set up the routing rules

**Reply by paulmiu** [reply]
@kangmingtay: This couldn't have been the issue, because the SMTP relay with the same credentials worked well everywhere else.

I've just tried all different combinations and interestingly all of them work well now for us without having changed anything in the codebase.

So, it's possible now to use the Google Workspace SMTP Relay with port 465 or 587 and a different sender email address than the one used for authentication.

It might have been fixed in the meantime.

### EDIT:

Well, I was a bit quick to judge.

What we found out is that it works for some Google Workspace accounts but not for others.
I'll add the required permissions and how to grant them here tomorrow.

**Reply by griiettner** [reply]
I'm having the issue where my app is unable to register users, because has trouble to send confirmation email.

Do you mind sharing the setup you used to make the setup works? Thanks in advance.

**Reply by deblanco** [reply]
@alexbaramilis, I can confirm that your solution with a non-admin account is working for me too
Thanks!

**Reply by nberney** [reply]
A good video that helped me how to Generate App Password in Gmail/Google Workspace: https://www.youtube.com/watch?v=lSURGX0JHbA

**Comment by groksrc** [comment]
Just an FYI from someone who has been working with GCP a lot. When you make changes to the SMTP configuration in the admin console the user interface won't tell you this but it is SLOW to take effect, sometimes up to 15 minutes or longer. If you've made all of the changes you think are appropriate, my recommendation is to just wait -- grab some coffee or whatever and then come back to it. This is not the only thing in Google that works this way, but it can be incredibly frustrating if you don't know to expect this behavior.

**Comment by jmordkoff** [comment]
If you have a personal gmail account, then the outbound email will ALWAYS have the from field set to your primary email address. 

But if you have an org, then you can set up an connection using smtp-relay.gmail.com, authenticate using your admin's email address and an app password and then you have flexibility on the from / replyto fields (based on the option you selected when you created the smtp-relay service in google as documented so well above), e.g. you can start using 'noreply@myorg.com' as the replyto address.

**Comment by mehdinathani** [comment]
I have free gmail account, can i use this?

**Comment by tamdao** [comment]
I follow this [guide](https://github.com/orgs/supabase/discussions/5320#discussioncomment-7836793) and got this error:
<img width="1219" alt="Screenshot 2024-05-28 at 04 56 22" src="https://github.com/supabase/supabase/assets/3086267/2f41db06-eda8-4d0e-9981-18dd28fe1233">

How can I fix this issue?

**Comment by ramijames** [comment]
If anyone is still encountering this (like I was) and needs some more clear directions, I put together a write-up of the solution here: https://www.ramijames.com/labs/supabase-email-authentication-with-google-workspace

It boils down to: 
1. Use SMTP Host: smtp-relay.gmail.com
2. Use SMTP Port: 465
3. Have 2fa enabled
4. Generate and use an app password
5. Use your admin account to authenticate with the above
